# Integrated Project: Maji Ndogo Part 2
© ExploreAI Academy

## Maji Ndogo: From analysis to action
# Clustering data to unveil Maji Ndogo's water crisis

> ⚠️ This notebook will not run on Google Colab because it cannot connect to a local database. Please make sure that this notebook is running on the same local machine as your MySQL Workbench installation and MySQL `mdd_water_services` database.




### Cleaning our data
Ok, bring up the employee table. It has info on all of our workers, but note that the email addresses have not been added. We will have to send them reports and figures, so let's update it. Luckily the emails for our department are easy: first_name.last_name@ndogowater.gov.

#### We can determine the email address for each employee by:
- selecting the employee_name column
- replacing the space with a full stop
- make it lowercase
- and stitch it all together

## Connecting to our MySQL database

Using our `Access_to_Basic_Services` table created in MySQL Workbench, we want to answer some questions on the range of our dataset. We can apply the same queries in MySQL Workbench and in this notebook if we connect to our MySQL server. Since we have a MySQL database, we can connect to it using mysql and pymysql.

In [37]:
# To display all row in ouputs
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

In [38]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook. 
# If you get an error here, make sure that mysql and pymysql are installed correctly. 

%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [41]:
# Establish a connection to the local database using the '%sql' magic command.
# Replace 'password' with our connection password and `db_name` with our database name. 
# If you get an error here, please make sure the database name or password is correct.

%sql mysql+pymysql://root:599190100@localhost:3306/md_water_services

Connecting and switching to connection 'mysql+pymysql://root:***@localhost:3306/md_water_services'

In [45]:
%%sql
Show tables;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

8 rows affected.

+-----------------------------+
| Tables_in_md_water_services |
+-----------------------------+
|       data_dictionary       |
|           employee          |
|     global_water_access     |
|           location          |
|            visits           |
|        water_quality        |
|         water_source        |
|        well_pollution       |
+-----------------------------+

In [47]:
%%sql

SELECT
    *
From
    data_dictionary;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

49 rows affected.

+---------------------+--------------------------+---------------------------------------------------------------------------------------------------------+--------------+-----------------------------+
|      table_name     |       column_name        |                                               description                                               |   datatype   |          related_to         |
+---------------------+--------------------------+---------------------------------------------------------------------------------------------------------+--------------+-----------------------------+
|       employee      |   assigned_employee_id   |                                   Unique ID assigned to each employee                                   |     INT      |            visits           |
|       employee      |      employee_name       |                                           Name of the employee                                          | VARCHAR(255) |                             |
|       employee      |       phone_number       |                                      Contact number of the employee                                     | VARCHAR(15)  |                             |
|       employee      |          email           |                                      Email address of the employee                                      | VARCHAR(255) |                             |
|       employee      |         address          |                                   Residential address of the employee                                   | VARCHAR(255) |                             |
|       employee      |        town_name         |                               Name of the town where the employee resides                               | VARCHAR(255) |                             |
|       employee      |      province_name       |                             Name of the province where the employee resides                             | VARCHAR(255) |                             |
|       employee      |         position         |                                  Position or job title of the employee                                  | VARCHAR(255) |                             |
|        visits       |        record_id         |                                     Unique ID assigned to each visit                                    |     int      | water_quality, water_source |
|        visits       |       location_id        |                                        ID of the location visited                                       | varchar(255) |           location          |
|        visits       |        source_id         |                                      ID of the water source visited                                     | varchar(510) |        well_pollution       |
|        visits       |      time_of_record      |                                        Date and time of the visit                                       |   datetime   |                             |
|        visits       |       visit_count        |                                  Number of visits made to this location                                 |     int      |                             |
|        visits       |      time_in_queue       |                    Time spent by people waiting for water in a queue at the location                    |     int      |                             |
|        visits       |   assigned_employee_id   |                               ID of the employee who visited the location                               |     int      |           employee          |
|    water_quality    |        record_id         |                                    Unique ID assigned to each record                                    |     int      |                             |
|    water_quality    | subjective_quality_score |                      Score representing the subjective quality of the water source                      |     i

In [77]:
%%sql
 
# Replace the space with a full stop AND Make it all lower case  PAGE (6)

Select
    LOWER(REPLACE(employee_name,' ','.'))
FROM
    employee
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+---------------------------------------+
| LOWER(REPLACE(employee_name,' ','.')) |
+---------------------------------------+
|              amara.jengo              |
|              bello.azibo              |
|              bakari.iniko             |
|             malachi.mavuso            |
|              cheche.buhle             |
|             zuriel.matembo            |
|              deka.osumare             |
|             lalitha.kaburi            |
|              enitan.zuri              |
|               farai.nia               |
+---------------------------------------+

In [79]:
%%sql

# Replace the space with a full stop AND Make it all lower case, and use CONCAT() '@ndogowater.gov') AS new_email −− add it all together   PAGE (6)

Select
    CONCAT(LOWER(REPLACE(employee_name,' ','.')), '@ndogowater.gov')
FROM
    employee
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+------------------------------------------------------------------+
| CONCAT(LOWER(REPLACE(employee_name,' ','.')), '@ndogowater.gov') |
+------------------------------------------------------------------+
|                    amara.jengo@ndogowater.gov                    |
|                    bello.azibo@ndogowater.gov                    |
|                   bakari.iniko@ndogowater.gov                    |
|                  malachi.mavuso@ndogowater.gov                   |
|                   cheche.buhle@ndogowater.gov                    |
|                  zuriel.matembo@ndogowater.gov                   |
|                   deka.osumare@ndogowater.gov                    |
|                  lalitha.kaburi@ndogowater.gov                   |
|                    enitan.zuri@ndogowater.gov                    |
|                     farai.nia@ndogowater.gov                     |
+------------------------------------------------------------------+

In [80]:
%%sql

# Use the employee table to count how many of our employees live in each town. 
# Think carefully about what function we should use and how we should aggregate the data.   PAGE (9)

SELECT DISTINCT
    province_name,
    town_name,
    COUNT(town_name) AS num_employees
From
    employee
GROUP BY
    1,2
ORDER BY
    1,3 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

15 rows affected.

+---------------+-----------+---------------+
| province_name | town_name | num_employees |
+---------------+-----------+---------------+
|     Akatsi    |   Rural   |       5       |
|     Akatsi    |   Lusaka  |       4       |
|     Akatsi    |   Harare  |       3       |
|     Akatsi    |  Kintampo |       1       |
|     Amanzi    |   Dahabu  |       6       |
|     Amanzi    |   Rural   |       1       |
|    Hawassa    |   Rural   |       7       |
|    Hawassa    |  Zanzibar |       4       |
|    Hawassa    |   Serowe  |       3       |
|    Hawassa    |  Yaounde  |       1       |
|    Kilimani   |   Rural   |       9       |
|    Kilimani   |   Harare  |       2       |
|    Kilimani   |   Ilanga  |       1       |
|     Sokoto    |   Rural   |       7       |
|     Sokoto    |   Ilanga  |       2       |
+---------------+-----------+---------------+

In [81]:
%%sql

# Let's first look at the number of records each employee collected. So find the correct table, 
# figure out what function to use and how to group, order and limit the results to only see the top 3 employee_ids 
# with the highest number of locations visited  PAGE (10)

SELECT
    assigned_employee_id,
    COUNT(visit_count) AS num_of_visits
FROM
    visits
GROUP BY
    1
ORDER BY
    2 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

29 rows affected.

+----------------------+---------------+
| assigned_employee_id | num_of_visits |
+----------------------+---------------+
|          1           |      3708     |
|          30          |      3676     |
|          34          |      3539     |
|          3           |      3420     |
|          10          |      3407     |
|          8           |      3351     |
|          5           |      3284     |
|          36          |      3249     |
|          48          |      2933     |
|          28          |      2762     |
|          12          |      2561     |
|          42          |      2496     |
|          40          |      2344     |
|          38          |      2121     |
|          2           |      2033     |
|          24          |      2015     |
|          32          |      1852     |
|          18          |      1749     |
|          46          |      1632     |
|          14          |      1441     |
|          7           |      1259     |
|          6           |      1182     |
|          4           |      1136     |
|          0           |      1099     |
|          16          |      802      |
|          26          |      719      |
|          44          |      218      |
|          22          |      143      |
|          20          |       15      |
+----------------------+---------------+

In [83]:
%%sql

# Create a query that counts the number of records per town  PAGE (11)

SELECT
    town_name,
    COUNT(town_name) AS records_per_town
FROM
    location
GROUP BY
    1
ORDER BY
    2 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

25 rows affected.

+-----------+------------------+
| town_name | records_per_town |
+-----------+------------------+
|   Rural   |      23740       |
|   Harare  |       1650       |
|   Amina   |       1090       |
|   Lusaka  |       1070       |
|   Mrembo  |       990        |
|   Asmara  |       930        |
|   Dahabu  |       930        |
|  Kintampo |       780        |
|   Ilanga  |       780        |
|  Isiqalo  |       770        |
|    Zuri   |       750        |
|   Amara   |       710        |
|  Majengo  |       620        |
|   Pwani   |       520        |
|    Kofi   |       510        |
|   Serowe  |       480        |
|   Bahari  |       470        |
|   Marang  |       460        |
|  Abidjan  |       400        |
|   Bello   |       400        |
|   Cheche  |       370        |
|    Deka   |       350        |
|   Djenne  |       310        |
|  Zanzibar |       300        |
|  Yaounde  |       270        |
+-----------+------------------+

In [84]:
%%sql

#  Now count the records per province.  PAGE (12)

SELECT
    province_name,
    COUNT(province_name) AS records_per_province
FROM
    location
GROUP BY
    1
ORDER BY
    2 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+---------------+----------------------+
| province_name | records_per_province |
+---------------+----------------------+
|    Kilimani   |         9510         |
|     Akatsi    |         8940         |
|     Sokoto    |         8220         |
|     Amanzi    |         6950         |
|    Hawassa    |         6030         |
+---------------+----------------------+

### Can you find a way to do the following:   PAGE (13)
-- 1. Create a result set showing:
• province_name
• town_name
• An aggregated count of records for each town (consider naming this records_per_town).
• Ensure your data is grouped by both province_name and town_name.
2. Order your results primarily by province_name. Within each province, further sort the towns by their record counts in descending order.

In [85]:
%%sql

SELECT
    province_name,
    town_name,
    COUNT(town_name) AS records_per_town
FROM
    location
GROUP BY
    1,2
ORDER BY
    1,3 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

31 rows affected.

+---------------+-----------+------------------+
| province_name | town_name | records_per_town |
+---------------+-----------+------------------+
|     Akatsi    |   Rural   |       6290       |
|     Akatsi    |   Lusaka  |       1070       |
|     Akatsi    |   Harare  |       800        |
|     Akatsi    |  Kintampo |       780        |
|     Amanzi    |   Rural   |       3100       |
|     Amanzi    |   Asmara  |       930        |
|     Amanzi    |   Dahabu  |       930        |
|     Amanzi    |   Amina   |       670        |
|     Amanzi    |   Pwani   |       520        |
|     Amanzi    |  Abidjan  |       400        |
|     Amanzi    |   Bello   |       400        |
|    Hawassa    |   Rural   |       3900       |
|    Hawassa    |   Serowe  |       480        |
|    Hawassa    |   Amina   |       420        |
|    Hawassa    |    Deka   |       350        |
|    Hawassa    |   Djenne  |       310        |
|    Hawassa    |  Zanzibar |       300        |
|    Hawassa    |  Yaounde  |       270        |
|    Kilimani   |   Rural   |       5440       |
|    Kilimani   |   Mrembo  |       990        |
|    Kilimani   |   Harare  |       850        |
|    Kilimani   |  Isiqalo  |       770        |
|    Kilimani   |    Zuri   |       750        |
|    Kilimani   |   Amara   |       710        |
|     Sokoto    |   Rural   |       5010       |
|     Sokoto    |   Ilanga  |       780        |
|     Sokoto    |  Majengo  |       620        |
|     Sokoto    |    Kofi   |       510        |
|     Sokoto    |   Bahari  |       470        |
|     Sokoto    |   Marang  |       460        |
|     Sokoto    |   Cheche  |       370        |
+---------------+-----------+------------------+

In [89]:
%%sql

# Finally, look at the number of records for each location type   PAGE (13)

SELECT
    location_type,
    COUNT(location_type) AS num_of_location_type
FROM
    location
GROUP BY
    1
ORDER BY
    1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2 rows affected.

+---------------+----------------------+
| location_type | num_of_location_type |
+---------------+----------------------+
|     Rural     |        23740         |
|     Urban     |        15910         |
+---------------+----------------------+

In [99]:
%%sql

# We can see that there are more rural sources than urban,
# but it's really hard to understand those numbers. Percentages are more relatable.
# If we use SQL as a very overpowered calculator:
SELECT
    location_type,
    COUNT(*) * 100 / (SELECT 
                          COUNT(*) 
                      FROM 
                          location) AS pct_of_location_type
FROM
    location
GROUP BY
    1
ORDER BY
    2 DESC;

# We can see that 60% of all water sources in the data set are in rural communities.

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2 rows affected.

+---------------+----------------------+
| location_type | pct_of_location_type |
+---------------+----------------------+
|     Rural     |       59.8739        |
|     Urban     |       40.1261        |
+---------------+----------------------+

So again, what are some of the insights we gained from the location table?
1. Our entire country was properly canvassed, and our dataset represents the situation on the ground.
2. 60% of our water sources are in rural communities across Maji Ndogo. We need to keep this in mind when we make decisions

In [101]:
%%sql

SELECT
    *
FROM
    water_source
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| AkHa00000224 |     tap_in_home      |           956           |
| AkHa00001224 |  tap_in_home_broken  |           930           |
| AkHa00002224 |  tap_in_home_broken  |           486           |
| AkHa00003224 |         well         |           364           |
| AkHa00004224 |  tap_in_home_broken  |           942           |
| AkHa00005224 |     tap_in_home      |           736           |
| AkHa00006224 |     tap_in_home      |           882           |
| AkHa00007224 |     tap_in_home      |           554           |
| AkHa00008224 |         well         |           398           |
| AkHa00009224 |         well         |           346           |
+--------------+----------------------+-------------------------+

### The way I look at this table; 
we have access to different water source types and the number of people using each source.
These are the questions that I am curious about.
1. How many people did we survey in total?
2. How many wells, taps and rivers are there?
3. How many people share particular types of water sources on average?
4. How many people are getting water from each type of source? 

In [102]:
%%sql

# 1. How many people did we survey in total?

SELECT
    SUM(number_of_people_served) AS total_number_of_people_served
FROM
    water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+-------------------------------+
| total_number_of_people_served |
+-------------------------------+
|            27628140           |
+-------------------------------+

In [103]:
%%sql
# 2. How many wells, taps and rivers are there? PAGE(15)

SELECT
    type_of_water_source,
    COUNT(type_of_water_source) AS num_of__water_sources
FROM
    water_source
GROUP BY
    1
ORDER BY
    2 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+-----------------------+
| type_of_water_source | num_of__water_sources |
+----------------------+-----------------------+
|         well         |         17383         |
|     tap_in_home      |          7265         |
|  tap_in_home_broken  |          5856         |
|      shared_tap      |          5767         |
|        river         |          3379         |
+----------------------+-----------------------+

In [106]:
%%sql
# 3. How many people share particular types of water sources on average?  PAGE(16)

SELECT
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) AS avg_number_of_people_served
FROM
    water_source
GROUP BY
    1
ORDER BY
    2 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+-------------------------------+
| type_of_water_source | total_number_of_people_served |
+----------------------+-------------------------------+
|      shared_tap      |              2071             |
|        river         |              699              |
|  tap_in_home_broken  |              649              |
|     tap_in_home      |              644              |
|         well         |              279              |
+----------------------+-------------------------------+

In [107]:
%%sql
# Now let’s calculate the total number of people served by each type of water source in total, to make it easier to interpret, order them so the most people served by a source is at the top.   PAGE(19)

SELECT
    type_of_water_source,
    ROUND(SUM(number_of_people_served)) AS total_number_of_people_served
FROM
    water_source
GROUP BY
    1
ORDER BY
    2 DESC


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+-----------------------------+
| type_of_water_source | avg_number_of_people_served |
+----------------------+-----------------------------+
|      shared_tap      |           11945272          |
|         well         |           4841724           |
|     tap_in_home      |           4678880           |
|  tap_in_home_broken  |           3799720           |
|        river         |           2362544           |
+----------------------+-----------------------------+

In [119]:
%%sql

# Next, calculate the percentages using the total we just got.  
# Let's round that off to 0 decimals, and order the results.   PAGE(21)

SELECT
    type_of_water_source,
    ROUND(SUM(number_of_people_served) * 100 /(SELECT
                                                   SUM(number_of_people_served)
                                                FROM water_source),0) AS pct_number_of_people_served
FROM
    water_source
GROUP BY
    1
ORDER BY
    2 DESC;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+-----------------------------+
| type_of_water_source | pct_number_of_people_served |
+----------------------+-----------------------------+
|      shared_tap      |              43             |
|         well         |              18             |
|     tap_in_home      |              17             |
|  tap_in_home_broken  |              14             |
|        river         |              9              |
+----------------------+-----------------------------+

In [136]:
%%sql

# So use a window function on the total people served column, converting it into a rank.    PAGE(23)
# But think about this: If someone has a tap in their home, 
# they already have the best source available. Since we can’t do anything more to improve
# this, we should remove tap_in_home from the ranking before we continue.

SELECT
    type_of_water_source,
    ROUND(SUM(number_of_people_served)) AS total_number_of_people_served,
    RANK() OVER (ORDER BY 
                     SUM(number_of_people_served) DESC) AS source_rank
FROM
    water_source
WHERE
    type_of_water_source != 'tap_in_home_broken'
GROUP BY
    1
ORDER BY
    3

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

+----------------------+-------------------------------+-------------+
| type_of_water_source | total_number_of_people_served | source_rank |
+----------------------+-------------------------------+-------------+
|      shared_tap      |            11945272           |      1      |
|         well         |            4841724            |      2      |
|     tap_in_home      |            4678880            |      3      |
|        river         |            2362544            |      4      |
+----------------------+-------------------------------+-------------+

In [143]:
%%sql

# So create a query to do this, and keep these requirements in mind:   PAGE(24)
# 1. The sources within each type should be assigned a rank.
# 2. Limit the results to only improvable sources.
# 3. Think about how to partition, filter and order the results set.
# 4. Order the results to see the top of the list.

SELECT
    source_id,
    type_of_water_source,
    number_of_people_served,
    DENSE_RANK() OVER (ORDER BY 
                     number_of_people_served DESC) AS source_rank
FROM
    water_source
WHERE
    type_of_water_source != 'tap_in_home_broken'
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+--------------+----------------------+-------------------------+-------------+
|  source_id   | type_of_water_source | number_of_people_served | source_rank |
+--------------+----------------------+-------------------------+-------------+
| HaRu19509224 |      shared_tap      |           3998          |      1      |
| AkRu05603224 |      shared_tap      |           3998          |      1      |
| AkRu04862224 |      shared_tap      |           3996          |      2      |
| KiHa22867224 |      shared_tap      |           3996          |      2      |
| AmAs10911224 |      shared_tap      |           3996          |      2      |
| HaRu19839224 |      shared_tap      |           3994          |      3      |
| KiZu31330224 |      shared_tap      |           3994          |      3      |
| KiZu31415224 |      shared_tap      |           3992          |      4      |
| KiRu28630224 |      shared_tap      |           3992          |      4      |
| KiRu26218224 |      shared_tap      |           3990          |      5      |
+--------------+----------------------+-------------------------+-------------+

### Ok, these are some of the things I think are worth looking at:
1. How long did the survey take?
2. What is the average total queue time for water?
3. What is the average queue time on different days?
4. How can we communicate this information efficiently?

In [154]:
%%sql

# Question 1:   PAGE(27)
# To calculate how long the survey took,
# we need to get the first and last dates (which functions can find the largest/smallest value), and subtract them. 
# Remember with DateTime data, we can't just subtract the values. We have to use a function to get the difference in days.

SELECT
    TIMESTAMPDIFF(DAY,MIN(time_of_record),MAX(time_of_record)) AS project_duration
FROM
    visits

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+------------------+
| project_duration |
+------------------+
|       924        |
+------------------+

In [161]:
%%sql
# Question 2:   PAGE(28)
# Let's see how long people have to queue on average in Maji Ndogo. 
# Keep in mind that many sources like taps_in_home have no queues. These are just recorded as 0 in the time_in_queue column, 
# so when we calculate averages, we need to exclude those rows. Try using NULLIF() do to this.

SELECT
    AVG(NULLIF(time_in_queue,0)) AS avg_time_in_queue
FROM
    visits

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+-------------------+
| avg_time_in_queue |
+-------------------+
|      123.2574     |
+-------------------+

In [164]:

%%sql
# So let's look at the queue times aggregated across the different days of the week.     PAGE(29)

SELECT
    DAYNAME(time_of_record) AS day_of_week,
    AVG(time_in_queue) AS avg_time_in_queue
FROM
    visits
GROUP BY
    1
ORDER BY
    2 DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

7 rows affected.

+-------------+-------------------+
| day_of_week | avg_time_in_queue |
+-------------+-------------------+
|   Saturday  |      246.2941     |
|    Sunday   |      81.5294      |
|    Monday   |      59.6780      |
|    Friday   |      52.7397      |
|   Tuesday   |      47.0795      |
|   Thursday  |      46.0164      |
|  Wednesday  |      42.5434      |
+-------------+-------------------+

In [183]:
%%sql
# Question 4:    PAGE(31)
# We can also look at what time during the day people collect water. 
# Try to order the results in a meaningful way.

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    AVG(time_in_queue) AS avg_time_in_queue
FROM
    visits
GROUP BY
    1
ORDER BY
    1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

+-------------+-------------------+
| hour_of_day | avg_time_in_queue |
+-------------+-------------------+
|    06:00    |      148.9306     |
|    07:00    |      149.1307     |
|    08:00    |      148.8944     |
|    09:00    |      49.2796      |
|    10:00    |      47.5738      |
|    11:00    |      45.9081      |
|    12:00    |      46.5188      |
|    13:00    |      47.0202      |
|    14:00    |      47.1823      |
|    15:00    |      47.6417      |
|    16:00    |      47.4669      |
|    17:00    |      148.7897     |
|    18:00    |      146.8198     |
|    19:00    |      167.7027     |
+-------------+-------------------+

In [194]:
%%sql
# To filter a row we use WHERE, but using CASE() in SELECT can filter columns. 
# We can use a CASE() function for each day to separate the queue time column into a column for each day. 
# Let’s begin by only focusing on Sunday. 
# So, when a row's DAYNAME(time_of_record) is Sunday, we make that value equal to time_in_queue, and NULL for any days.

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    DAYNAME(time_of_record) AS DAY_NAME,
    CASE
        WHEN DAYNAME(time_of_record) = 'Saturday' 
            THEN time_in_queue
        WHEN DAYNAME(time_of_record) = 'Sunday' 
            THEN time_in_queue
        WHEN DAYNAME(time_of_record) = 'Monday' 
            THEN time_in_queue
        WHEN DAYNAME(time_of_record) = 'Tuesday' 
            THEN time_in_queue
        WHEN DAYNAME(time_of_record) = 'Wednesday' 
            THEN time_in_queue
        WHEN DAYNAME(time_of_record) = 'Thursday' 
            THEN time_in_queue
        WHEN DAYNAME(time_of_record) = 'Friday' 
            THEN time_in_queue
        ELSE NULL
    END AS Time_of_waiting
FROM
    visits
WHERE
    time_in_queue != 0
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+-------------+----------+-----------------+
| hour_of_day | DAY_NAME | Time_of_waiting |
+-------------+----------+-----------------+
|    09:00    |  Friday  |        15       |
|    09:00    |  Friday  |        62       |
|    10:00    |  Friday  |        28       |
|    10:00    |  Friday  |        9        |
|    11:00    |  Friday  |        17       |
|    11:00    |  Friday  |       240       |
|    12:00    |  Friday  |        20       |
|    12:00    |  Friday  |       171       |
|    13:00    |  Friday  |        28       |
|    13:00    |  Friday  |        16       |
+-------------+----------+-----------------+

In [207]:
%%sql
# creating a pivot table of time waiting in each day   PAGE(35)

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
-- Sunday
    ROUND(AVG(
    	CASE
    		WHEN DAYNAME(time_of_record) = 'Sunday'
                  THEN time_in_queue
           ELSE NULL
        END),0) AS Sunday,
    -- Monday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Monday' 
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Monday,
    -- Tuesday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Tuesday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Tuesday,
    -- Wednesday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Wednesday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Wednesday,
    -- Thursday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Thursday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Thursday,
    -- Friday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Friday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Friday,
    -- Saturday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Saturday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Saturday
FROM
   visits
WHERE
   time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
     1
ORDER BY
     1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

+-------------+--------+--------+---------+-----------+----------+--------+----------+
| hour_of_day | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
|    06:00    |   79   |  190   |   134   |    112    |   134    |  153   |   247    |
|    07:00    |   82   |  186   |   128   |    111    |   139    |  156   |   247    |
|    08:00    |   86   |  183   |   130   |    119    |   129    |  153   |   247    |
|    09:00    |   84   |  127   |   105   |     94    |    99    |  107   |   252    |
|    10:00    |   83   |  119   |    99   |     89    |    95    |  112   |   259    |
|    11:00    |   78   |  115   |   102   |     86    |    99    |  104   |   236    |
|    12:00    |   78   |  115   |    97   |     88    |    96    |  109   |   239    |
|    13:00    |   81   |  122   |    97   |     98    |   101    |  115   |   242    |
|    14:00    |   83   |  127   |   104   |     92    |    96    |  110   |   244    |
|    15:00    |   83   |  126   |   104   |     88    |    92    |  110   |   248    |
|    16:00    |   83   |  127   |    99   |     90    |    99    |  109   |   251    |
|    17:00    |   79   |  181   |   135   |    121    |   129    |  151   |   251    |
|    18:00    |   80   |  174   |   122   |    113    |   132    |  158   |   240    |
|    19:00    |  127   |  159   |   145   |    176    |   137    |  103   |   282    |
+-------------+--------+--------+---------+-----------+----------+--------+----------+

# Questions of Integrated Project: Maji Ndogo Part 2

In [209]:
%%sql

# Q1
# Which SQL query will produce the date format "DD Month YYYY"
# from the time_of_record column in the visits table,
# as a single column? Note: Monthname() acts in a similar way to DAYNAME().


SELECT 
    CONCAT(day(time_of_record), " ", monthname(time_of_record), " ", year(time_of_record)) AS Standard_date
FROM 
   visits
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+----------------+
| Standard_date  |
+----------------+
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
| 1 January 2021 |
+----------------+

### Q2

You are working with an SQL query designed 

to calculate the Annual Rate of Change (ARC) for basic rural water services:

SELECT
    name,
    wat_bas_r - LAG(wat_bas_r) OVER (PARTITION BY (a) ORDER BY (b)) 
FROM 
    global_water_access
ORDER BY
     name;

To accomplish this task, what should you use for placeholders (a) and (b)?


In [212]:
%%sql
# Q2
SELECT
    name,
    wat_bas_r - LAG(wat_bas_r) OVER (PARTITION BY name ORDER BY year) AS ARC 
FROM 
    global_water_access
ORDER BY
     name
LIMIT 10

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+----------------+--------------------+
|      name      |        ARC         |
+----------------+--------------------+
|  Afghanistan   |        None        |
|  Afghanistan   | 13.339405059814453 |
|    Albania     |        None        |
|    Albania     | 3.4641036987304688 |
|    Algeria     |        None        |
|    Algeria     | 1.6847991943359375 |
| American Samoa |        None        |
| American Samoa |        None        |
|    Andorra     |        None        |
|    Andorra     |        0.0         |
+----------------+--------------------+

### Q3
What are the names of the two worst-performing employees who visited the fewest sites,

and how many sites did the worst-performing employee visit? 

Modify your queries from the “Honouring the workers” section


In [221]:
%%sql
# Q3

SELECT
    assigned_employee_id,
    SUM(visit_count) AS num_of_visit_count
FROM
    visits
GROUP BY
    1
ORDER BY
    2
LIMIT 2

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2 rows affected.

+----------------------+--------------------+
| assigned_employee_id | num_of_visit_count |
+----------------------+--------------------+
|          20          |         43         |
|          22          |        339         |
+----------------------+--------------------+

In [226]:
%%sql

SELECT
    *
FROM
    employee
WHERE
    assigned_employee_id = '20'
    OR assigned_employee_id = '22'

------------------------------------> ANSWER: Lesedi Kofi, Kunto Asha (15)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2 rows affected.

+----------------------+---------------+---------------+-------+------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name |  phone_number | email |        address         | province_name | town_name |    position    |
+----------------------+---------------+---------------+-------+------------------------+---------------+-----------+----------------+
|          20          |   Kunto Asha  | +99176320477  |  None | 30 Nyoka Achebe Street |     Amanzi    |   Dahabu  | Field Surveyor |
|          22          |  Lesedi Kofi  | +99611183730  |  None |    52 Moroni Avenue    |     Sokoto    |   Rural   | Field Surveyor |
+----------------------+---------------+---------------+-------+------------------------+---------------+-----------+----------------+

In [234]:
%%sql

# Q4
# What does the following query do?

SELECT 
    location_id,
    time_in_queue,
    visit_count,
    AVG(time_in_queue) OVER (PARTITION BY location_id ORDER BY visit_count) AS total_avg_queue_time
FROM 
    visits
WHERE 
     visit_count > 1
ORDER BY 
    location_id, 
    time_of_record
LIMIT 10

--------------> ANSWER: It computes an average queue time for shared taps visited more than once, which is updated each time a source is visited.

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

10 rows affected.

+-------------+---------------+-------------+----------------------+
| location_id | time_in_queue | visit_count | total_avg_queue_time |
+-------------+---------------+-------------+----------------------+
|  AkHa00036  |      113      |      2      |       113.0000       |
|  AkHa00036  |      229      |      3      |       171.0000       |
|  AkHa00036  |      400      |      4      |       247.3333       |
|  AkHa00036  |      120      |      5      |       215.5000       |
|  AkHa00036  |      185      |      6      |       209.4000       |
|  AkHa00036  |      158      |      7      |       200.8333       |
|  AkHa00036  |      226      |      8      |       204.4286       |
|  AkHa00090  |       50      |      2      |       50.0000        |
|  AkHa00090  |       63      |      3      |       56.5000        |
|  AkHa00090  |       12      |      4      |       41.6667        |
+-------------+---------------+-------------+----------------------+

In [238]:
%%sql

# Q5
# One of our employees, Farai Nia,
# lives at 33 Angelique Kidjo Avenue. What would be the result if we TRIM() her address?
# TRIM('33 Angelique Kidjo Avenue  ')

SELECT 
 TRIM('33 Angelique Kidjo Avenue  ') AS trim_result

-- ------------------------------------> ANSWER: ‘33 Angelique Kidjo Avenue’

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+---------------------------+
|        trim_result        |
+---------------------------+
| 33 Angelique Kidjo Avenue |
+---------------------------+

In [239]:
%%sql

# Q6
# How many employees live in Dahabu? 
# Rely on one of the queries we used in the project to answer this.

SELECT
    COUNT(*) AS num_of_Dahabu_employees
FROM
    employee
where
    town_name = 'Dahabu'
--------------------------------------> ANSWER: 6

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+-------------------------+
| num_of_Dahabu_employees |
+-------------------------+
|            6            |
+-------------------------+

In [245]:
%%sql

# Q7
# How many employees live in Harare, Kilimani?
# Modify one of your queries from the project to answer this question.

SELECT
    COUNT(*) AS num_of_employees
FROM
    employee
where
    province_name = 'Kilimani'
    AND town_name = 'Harare'

-- ------------------------------------> ANSWER: 2

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+------------------+
| num_of_employees |
+------------------+
|        2         |
+------------------+

In [259]:
%%sql

# Q8
# How many people share a well on average? Round your answer to 0 decimals.

SELECT
    ROUND(AVG(number_of_people_served),0)
FROM
    water_source
WHERE
    type_of_water_source = 'well'

--------------------------------------> ANSWER: 279

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+---------------------------------------+
| ROUND(AVG(number_of_people_served),0) |
+---------------------------------------+
|                  279                  |
+---------------------------------------+

### Q9

Consider the query we used to calculate the total number of people served:

SELECT
    SUM(number_of_people_served) AS population_served
FROM
   water_source
ORDER BY
    population_served

Which of the following lines of code will calculate the total number of people using some sort of tap?

### -- ------------------------------------> ANSWER: WHERE type_of_water_source LIKE "%tap%"

#### Below is Justifying my Answer

In [263]:
%%sql

# Q9 

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS population_served
FROM
   water_source
WHERE
    type_of_water_source LIKE '%tap%'
GROUP BY
    1
ORDER BY
    2 DESC


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

3 rows affected.

+----------------------+-------------------+
| type_of_water_source | population_served |
+----------------------+-------------------+
|      shared_tap      |      11945272     |
|     tap_in_home      |      4678880      |
|  tap_in_home_broken  |      3799720      |
+----------------------+-------------------+

### Q10

Use the pivot table we created to answer the following question. What are the average queue times for the following times?

Saturday from 12:00 to 13:00
Tuesday from 18:00 to 19:00
Sunday from 09:00 to 10:00



### -- ------------------------------------> ANSWER: Saturday: 239, Tuesday: 122, Sunday: 84

#### Below is Justifying my Answer

In [264]:
%%sql

# Q10

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
-- Sunday
    ROUND(AVG(
    	CASE
    		WHEN DAYNAME(time_of_record) = 'Sunday'
                  THEN time_in_queue
           ELSE NULL
        END),0) AS Sunday,
    -- Monday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Monday' 
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Monday,
    -- Tuesday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Tuesday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Tuesday,
    -- Wednesday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Wednesday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Wednesday,
    -- Thursday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Thursday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Thursday,
    -- Friday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Friday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Friday,
    -- Saturday
    ROUND(AVG(
    	CASE
           WHEN DAYNAME(time_of_record) = 'Saturday'
                 THEN time_in_queue
           ELSE NULL
    	END),0) AS Saturday
FROM
   visits
WHERE
   time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
     1
ORDER BY
     1

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

+-------------+--------+--------+---------+-----------+----------+--------+----------+
| hour_of_day | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
|    06:00    |   79   |  190   |   134   |    112    |   134    |  153   |   247    |
|    07:00    |   82   |  186   |   128   |    111    |   139    |  156   |   247    |
|    08:00    |   86   |  183   |   130   |    119    |   129    |  153   |   247    |
|    09:00    |   84   |  127   |   105   |     94    |    99    |  107   |   252    |
|    10:00    |   83   |  119   |    99   |     89    |    95    |  112   |   259    |
|    11:00    |   78   |  115   |   102   |     86    |    99    |  104   |   236    |
|    12:00    |   78   |  115   |    97   |     88    |    96    |  109   |   239    |
|    13:00    |   81   |  122   |    97   |     98    |   101    |  115   |   242    |
|    14:00    |   83   |  127   |   104   |     92    |    96    |  110   |   244    |
|    15:00    |   83   |  126   |   104   |     88    |    92    |  110   |   248    |
|    16:00    |   83   |  127   |    99   |     90    |    99    |  109   |   251    |
|    17:00    |   79   |  181   |   135   |    121    |   129    |  151   |   251    |
|    18:00    |   80   |  174   |   122   |    113    |   132    |  158   |   240    |
|    19:00    |  127   |  159   |   145   |    176    |   137    |  103   |   282    |
+-------------+--------+--------+---------+-----------+----------+--------+----------+

#  
<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/EAI_Blue_Dark.png"  style="width:200px";/>
</div>